src/rgw/rgw_notify.cc

```diff
@@ -2,17 +2,530 @@
+            has_error = true;
+            return;
+          }
+          try {
+            // TODO move endpoint creation outside the entries loop
+            const auto push_endpoint = RGWPubSubEndpoint::create(record_with_endpoint.push_endpoint, record_with_endpoint.arn_topic,
+                RGWHTTPArgs(record_with_endpoint.push_endpoint_args), 
+                cct);
+            ldout(cct, 20) << "INFO: push endpoint created: " << record_with_endpoint.push_endpoint <<
+              " for entry: " << entry_idx << "/" << total_entries << " (marker: " << entry.marker << ")" << dendl;
+            const auto ret = push_endpoint->send_to_completion_async(cct, record_with_endpoint.record, optional_yield(io_context, yield));
+            if (ret < 0) {
+              ldout(cct, 5) << "WARNING: push entry: " << entry_idx << "/" << total_entries << " (marker: " << entry.marker << ") to endpoint: " << record_with_endpoint.push_endpoint 
+                << " failed. error: " << ret << " (will retry)" << dendl;
+              if (set_min_marker(end_marker, entry.marker) < 0) {
+                ldout(cct, 1) << "ERROR: cannot determin minimum between malformed markers: " << end_marker << ", " << entry.marker << dendl;
+              } else {
+                ldout(cct, 20) << "INFO: end marker for removal: " << end_marker << dendl;
+              }
+              has_error = true;
+              return;
+            } else {
+              ldout(cct, 20) << "INFO: push entry: " << entry_idx << "/" << total_entries << " (marker: " << entry.marker << ") to endpoint: " << record_with_endpoint.push_endpoint 
+                << " OK" <<  dendl;
+              if (perfcounter) perfcounter->inc(l_rgw_pubsub_push_ok);
+              // there is at least one entry to remove (was successfully published)
+              remove_entries = true;
+            }
+          } catch (const RGWPubSubEndpoint::configuration_error& e) {
+            ldout(cct, 5) << "WARNING: failed to create push endpoint: " 
+                << record_with_endpoint.push_endpoint << ". error: " << e.what() << " (will retry) " << dendl;
+            if (set_min_marker(end_marker, entry.marker) < 0) {
+              ldout(cct, 1) << "ERROR: cannot determin minimum between malformed markers: " << end_marker << ", " << entry.marker << dendl;
+            } else {
+              ldout(cct, 20) << "INFO: end marker for removal: " << end_marker << dendl;
+            }
+            has_error = true;
+            return;
+          }
+        });
+        ++entry_idx;
+      }
+
+      // wait for all pending work to finish
+      waiter.async_wait(yield);
+
+      // delete all published entries from queue
+      if (remove_entries) {
+        librados::ObjectWriteOperation op;
+        cls_2pc_queue_remove_entries(op, end_marker); 
+        // TODO call operate async with yield
+        const auto ret = rados_ioctx.operate(queue_name, &op);
```

Comment:
you might even want to fire this off asynchronously (without suspending or checking the result), so you can start processing in the next batch of entries immediately

Reply:
but then the entries may still be in the queue the next time i read from it?

Reply:
currently I'm just yielding on the operations